File: src/main/java/org/taktik/icure/logic/impl/MessageLogicImpl.java
/*
 * Copyright (C) 2018 Taktik SA
 *
 * This file is part of iCureBackend.
 *
 * iCureBackend is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as published by
 * the Free Software Foundation.
 *
 * iCureBackend is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with iCureBackend.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.taktik.icure.logic.impl;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.access.AccessDeniedException;
import org.springframework.stereotype.Service;
import org.taktik.icure.dao.DocumentDAO;
import org.taktik.icure.dao.MessageDAO;
import org.taktik.icure.dao.Option;
import org.taktik.icure.db.PaginatedList;
import org.taktik.icure.db.PaginationOffset;
import org.taktik.icure.entities.Message;
import org.taktik.icure.entities.User;
import org.taktik.icure.entities.embed.Delegation;
import org.taktik.icure.exceptions.CreationException;
import org.taktik.icure.exceptions.MissingRequirementsException;
import org.taktik.icure.logic.ICureSessionLogic;
import org.taktik.icure.logic.MessageLogic;
import org.taktik.icure.validation.aspect.Check;

import javax.security.auth.login.LoginException;
import javax.validation.constraints.NotNull;
import java.util.*;
import java.util.stream.Collectors;

@Service
public class MessageLogicImpl extends GenericLogicImpl<Message, MessageDAO> implements MessageLogic {
	private static Logger logger = LoggerFactory.getLogger(MessageLogicImpl.class);

	private DocumentDAO documentDAO;
	private MessageDAO messageDAO;
	private ICureSessionLogic sessionLogic;

	@Override
	public List<Message> listMessagesByHCPartySecretPatientKeys(List<String> secretPatientKeys) throws LoginException {
		User loggedUser = getLoggedHealthCarePartyUser();
		return messageDAO.findByHcPartyPatient(loggedUser.getHealthcarePartyId(), secretPatientKeys);
	}

	@Override
	public List<Message> setStatus(List<String> messageIds, int status) {
		return messageDAO.save(messageDAO.getList(messageIds).stream().map(m -> {
			m.setStatus(m.getStatus() != null ? (m.getStatus() | status) : status); return m;
		}).collect(Collectors.toList()));
	}


	@Override
	public PaginatedList<Message> findForCurrentHcParty(PaginationOffset paginationOffset) throws LoginException {
		User loggedUser = getLoggedHealthCarePartyUser();
		return messageDAO.findByHcParty(loggedUser.getHealthcarePartyId(), paginationOffset);
	}

	@Override
	public PaginatedList<Message> findByFromAddress(String partyId, String fromAddress, PaginationOffset<List<Object>> paginationOffset) {
		return messageDAO.findByFromAddress(partyId,fromAddress,paginationOffset);
	}

	@Override
	public PaginatedList<Message> findByToAddress(String partyId, String toAddress, PaginationOffset<List<Object>> paginationOffset, Boolean reverse) {
		return messageDAO.findByToAddress(partyId, toAddress, paginationOffset, reverse);
	}

	@Override
	public PaginatedList<Message> findByTransportGuidReceived(String partyId, String transportGuid, PaginationOffset<List<Object>> paginationOffset) {
		return messageDAO.findByTransportGuidReceived(partyId, transportGuid, paginationOffset);
	}

	@Override
	public PaginatedList<Message> findByTransportGuid(String partyId, String transportGuid, PaginationOffset<List<Object>> paginationOffset) {
		return messageDAO.findByTransportGuid(partyId, transportGuid, paginationOffset);
	}

	@Override
	public PaginatedList<Message> findByTransportGuidSentDate(String partyId, String transportGuid, Long fromDate, Long toDate, PaginationOffset<List<Object>> paginationOffset) {
		return messageDAO.findByTransportGuidSentDate(partyId, transportGuid, fromDate, toDate, paginationOffset);
	}

	@Override
	public Message addDelegation(String messageId, Delegation delegation) {
		Message message = messageDAO.get(messageId);
		message.addDelegation(delegation.getDelegatedTo(), delegation);
		return messageDAO.save(message);
	}

	@Override
	public Message addDelegations(String messageId, List<Delegation> delegations) {
		Message message = messageDAO.get(messageId);
		delegations.forEach(d->message.addDelegation(d.getDelegatedTo(),d));
		return messageDAO.save(message);
	}

	@Override
	public List<Message> getChildren(String messageId) {
		return messageDAO.getChildren(messageId);
	}

	@Override
	public List<List<Message>> getChildren(List<String> parentIds) {
		return messageDAO.getChildren(parentIds);
	}

	@Override
	public List<Message> getByTransportGuids(String hcpId, Set<String> transportGuids) {
		return messageDAO.getByTransportGuids(hcpId, transportGuids);
	}

	@Override
	public List<Message> listMessagesByInvoiceIds(List<String> ids) {
		return messageDAO.getByInvoiceIds(new HashSet<>(ids));
	}

	@Override
	public List<Message> listMessagesByExternalRefs(String hcPartyId, List<String> externalRefs) {
		return messageDAO.getByExternalRefs(hcPartyId, new HashSet<String>(externalRefs));
	}

	@Override
	public boolean createEntities(Collection<Message> entities, Collection<Message> createdEntities) throws Exception {
		User loggedUser = getLoggedHealthCarePartyUser();
		boolean success = true;
		for (Message message : entities) {
			if (message == null) {
				logger.error("Cannot create 'null' message. ");
			} else {
				if (message.getFromAddress()==null) { message.setFromAddress(loggedUser.getEmail()); }
				if (message.getFromHealthcarePartyId()==null) { message.setFromHealthcarePartyId(loggedUser.getHealthcarePartyId()); }

				success = success && super.createEntities(Collections.singletonList(message), createdEntities);
			}
		}
		return success;
	}

	@Override
	public Message createMessage(@Check @NotNull Message message) throws CreationException, LoginException {
		if (message == null) {
			throw new CreationException("Cannot create 'null' message. ");
		}

		List<Message> createdMessages = new ArrayList<>(1);
		try {
			createEntities(Collections.singleton(message), createdMessages);

		} catch (Exception e) {
			throw new CreationException("Could not create message. ", e);
		}
		return createdMessages.size() > 0 ? createdMessages.get(0) : null;
	}

	@Override
	public Message get(String messageId) {
		return messageDAO.get(messageId);
	}

	@Override
	public void modifyMessage(@Check @NotNull Message message) throws MissingRequirementsException {
		messageDAO.save(message);
	}

	@Override
	public void solveConflicts() {
		List<Message> messagesInConflict = messageDAO.listConflicts().stream().map(it -> messageDAO.get(it.getId(), Option.CONFLICTS)).collect(Collectors.toList());

		messagesInConflict.forEach(msg -> {
			Arrays.stream(msg.getConflicts()).map(c -> messageDAO.get(msg.getId(), c)).forEach(cp -> {
				msg.solveConflictWith(cp);
				messageDAO.purge(cp);
			});
			messageDAO.save(msg);
		});
	}


	@Autowired
	public void setDocumentDAO(DocumentDAO documentDAO) {
		this.documentDAO = documentDAO;
	}

	@Autowired
	public void setMessageDAO(MessageDAO messageDAO) {
		this.messageDAO = messageDAO;
	}

	@Autowired
	public void setSessionLogic(ICureSessionLogic sessionLogic) {
		this.sessionLogic = sessionLogic;
	}

	@Override
	protected MessageDAO getGenericDAO() {
		return messageDAO;
	}

	private User getLoggedHealthCarePartyUser() throws LoginException {
		User user = sessionLogic.getCurrentSessionContext().getUser();
		if (user == null || user.getHealthcarePartyId() == null) {
			throw new LoginException("You must be logged to perform this action");
		}
		return user;
	}


}

File: src/main/kotlin/org/taktik/icure/asynclogic/impl/MessageLogicImpl.kt
/*
 *  iCure Data Stack. Copyright (c) 2020 Taktik SA
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Affero General Public License as
 *     published by the Free Software Foundation, either version 3 of the
 *     License, or (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful, but
 *     WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *     Affero General Public License for more details.
 *
 *     You should have received a copy of the GNU Affero General Public
 *     License along with this program.  If not, see
 *     <https://www.gnu.org/licenses/>.
 */
package org.taktik.icure.asynclogic.impl

import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.flow.*
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import org.taktik.couchdb.ViewQueryResultEvent
import org.taktik.couchdb.entity.Option
import org.taktik.icure.asyncdao.DocumentDAO
import org.taktik.icure.asyncdao.MessageDAO
import org.taktik.icure.asynclogic.AsyncSessionLogic
import org.taktik.icure.asynclogic.MessageLogic
import org.taktik.icure.db.PaginationOffset
import org.taktik.icure.entities.Message
import org.taktik.icure.entities.embed.Delegation
import org.taktik.icure.entities.embed.MessageReadStatus
import org.taktik.icure.exceptions.CreationException
import org.taktik.icure.exceptions.MissingRequirementsException
import org.taktik.icure.exceptions.PersistenceException
import org.taktik.icure.utils.firstOrNull
import javax.security.auth.login.LoginException

@ExperimentalCoroutinesApi
@Service
class MessageLogicImpl(private val documentDAO: DocumentDAO, private val messageDAO: MessageDAO, private val sessionLogic: AsyncSessionLogic) : GenericLogicImpl<Message, MessageDAO>(sessionLogic), MessageLogic {

    @Throws(LoginException::class)
    override fun listMessagesByHCPartySecretPatientKeys(secretPatientKeys: List<String>) = flow<Message> {
        emitAll(messageDAO.findByHcPartyPatient(currentHealthCarPartyId(), secretPatientKeys))
    }

    @Throws(PersistenceException::class)
    override fun setStatus(messageIds: List<String>, status: Int) = flow<Message> {
        emitAll(messageDAO.save(messageDAO.getList(messageIds)
                .map {
                    it.copy(status = status or (it.status ?: 0))
                }.toList()))
    }

    @Throws(PersistenceException::class)
    override fun setReadStatus(messageIds: List<String>, userId: String, status: Boolean, time: Long) = flow<Message> {
        emitAll(
            messageDAO.save(messageDAO.getList(messageIds).map { m: Message ->
                if ((m.readStatus[userId]?.time ?: 0) < time) m.copy(readStatus = m.readStatus + (userId to MessageReadStatus(
                        read = status,
                        time = time
                ))) else m
            }.toList())
        )
    }

    @Throws(LoginException::class)
    override fun findForCurrentHcParty(paginationOffset: PaginationOffset<List<Any>>) = flow<ViewQueryResultEvent> {
        emitAll(messageDAO.findByHcParty(currentHealthCarPartyId(), paginationOffset))
    }

    override fun findMessagesByFromAddress(partyId: String, fromAddress: String, paginationOffset: PaginationOffset<List<Any>>) = flow<ViewQueryResultEvent> {
        emitAll(messageDAO.findByFromAddress(partyId, fromAddress, paginationOffset))
    }

    override fun findMessagesByToAddress(partyId: String, toAddress: String, paginationOffset: PaginationOffset<List<Any>>, reverse: Boolean?) = flow<ViewQueryResultEvent> {
        emitAll(messageDAO.findByToAddress(partyId, toAddress, paginationOffset, reverse))
    }

    override fun findMessagesByTransportGuidReceived(partyId: String, transportGuid: String?, paginationOffset: PaginationOffset<List<Any>>) = flow<ViewQueryResultEvent> {
        emitAll(messageDAO.findByTransportGuidReceived(partyId, transportGuid, paginationOffset))
    }

    override fun findMessagesByTransportGuid(partyId: String, transportGuid: String?, paginationOffset: PaginationOffset<List<Any>>) = flow<ViewQueryResultEvent> {
        emitAll(messageDAO.findByTransportGuid(partyId, transportGuid, paginationOffset))
    }

    override fun findMessagesByTransportGuidSentDate(partyId: String, transportGuid: String, fromDate: Long, toDate: Long, paginationOffset: PaginationOffset<List<Any>>) = flow<ViewQueryResultEvent> {
        emitAll(messageDAO.findByTransportGuidSentDate(partyId, transportGuid, fromDate, toDate, paginationOffset))
    }

    override suspend fun addDelegation(messageId: String, delegation: Delegation): Message? {
        val message = messageDAO.get(messageId)
        return delegation.delegatedTo?.let { healthcarePartyId ->
            message?.let { c -> messageDAO.save(c.copy(delegations = c.delegations + mapOf(
                    healthcarePartyId to setOf(delegation)
            )))}
        } ?: message
    }

    override suspend fun addDelegations(messageId: String, delegations: List<Delegation>): Message? {
        val message = messageDAO.get(messageId)
        return message?.let {
            return messageDAO.save(it.copy(
                    delegations = it.delegations +
                            delegations.mapNotNull { d -> d.delegatedTo?.let { delegateTo -> delegateTo to setOf(d) } }
            ))
        }
    }

    override fun getMessageChildren(messageId: String) = flow<Message> {
        emitAll(messageDAO.getChildren(messageId))
    }

    override fun getMessagesChildren(parentIds: List<String>) = flow<List<Message>> {
        emitAll(messageDAO.getChildren(parentIds))
    }

    override fun getMessagesByTransportGuids(hcpId: String, transportGuids: Set<String>) = flow<Message> {
        emitAll(messageDAO.getByTransportGuids(hcpId, transportGuids))
    }

    override fun listMessagesByInvoiceIds(ids: List<String>) = flow<Message> {
        emitAll(messageDAO.getByInvoiceIds(ids.toSet()))
    }

    override fun listMessagesByExternalRefs(hcPartyId: String, externalRefs: List<String>) = flow<Message> {
        emitAll(messageDAO.getByExternalRefs(hcPartyId, externalRefs.toSet()))
    }

    fun createMessages(entities: Collection<Message>, createdEntities: Collection<Message>) = flow {
        val loggedUser = sessionLogic.getCurrentSessionContext().getUser()

        emitAll(super.createEntities(entities.map {
            if (it.fromAddress == null || it.fromHealthcarePartyId == null)
                it.copy(
                        fromAddress = it.fromAddress ?: loggedUser.email,
                        fromHealthcarePartyId = it.fromHealthcarePartyId ?: loggedUser.healthcarePartyId
                )
            else it
        }))
    }

    @Throws(CreationException::class, LoginException::class)
    override suspend fun createMessage(message: Message) = fix(message) { message ->
        val createdMessages: List<Message> = ArrayList(1)
        createMessages(setOf(message), createdMessages).firstOrNull()
    }

    override suspend fun getMessage(messageId: String): Message? {
        return messageDAO.get(messageId)
    }

    @Throws(MissingRequirementsException::class)
    override suspend fun modifyMessage(message: Message) = fix(message) { message ->
        messageDAO.save(message)
    }

    override fun solveConflicts(): Flow<Message> =
            messageDAO.listConflicts().mapNotNull { messageDAO.get(it.id, Option.CONFLICTS)?.let { message ->
                message.conflicts?.mapNotNull { conflictingRevision -> messageDAO.get(message.id, conflictingRevision) }
                        ?.fold(message) { kept, conflict -> kept.merge(conflict).also { messageDAO.purge(conflict) } }
                        ?.let { mergedMessage -> messageDAO.save(mergedMessage) }
            } }

    override fun getGenericDAO(): MessageDAO {
        return messageDAO
    }

    private suspend fun currentHealthCarPartyId(): String = sessionLogic.getCurrentHealthcarePartyId()
            ?: throw LoginException("You must be logged to perform this action")

    companion object {
        private val logger = LoggerFactory.getLogger(MessageLogicImpl::class.java)
    }
}
